cilium/cmd/monitor.go

```diff
@@ -327,34 +404,29 @@ func runMonitor(args []string) {
 		}
 	}
 	fmt.Printf("Press Ctrl-C to quit\n")
-start:
-	conn, err := net.Dial("unix", defaults.MonitorSockPath)
-	if err != nil {
-		fmt.Printf("Error: unable to connect to monitor %s\n", err)
-		os.Exit(1)
-	}
-
-	defer conn.Close()
 
-	var meta payload.Meta
-	var pl payload.Payload
-	for {
-		if err := payload.ReadMetaPayload(conn, &meta, &pl); err != nil {
-			if err == io.EOF || err == io.ErrUnexpectedEOF {
-				// EOF may be due to invalid payload size. Close the connection just in case.
-				conn.Close()
-				log.WithError(err).Warn("connection closed")
-				time.Sleep(connTimeout)
-				goto start
-			} else {
-				log.WithError(err).Fatal("decoding error")
-			}
+	// On EOF, retry
+	// On other errors, exit
+	// always wait connTimeout when retrying
+	for ; ; time.Sleep(connTimeout) {
+		conn, version, err := openMonitorSock()
+		if err != nil {
+			log.WithError(err).Error("Cannot open monitor socket")
+			return
 		}
 
-		if pl.Type == payload.EventSample {
-			receiveEvent(pl.Data, pl.CPU)
-		} else /* if pl.Type == payload.RecordLost */ {
-			lostEvent(pl.Lost, pl.CPU)
+		err = consumeMonitorEvents(conn, version)
+		switch {
+		case err == nil:
+		// no-op
+
+		case err == io.EOF, err == io.ErrUnexpectedEOF:
+			log.WithError(err).Warn("connection closed")
+			time.Sleep(connTimeout)
```

Comment:
We seem to be sleeping twice now, once here, once in the for loop

Reply:
oops, yes, the outer loop now does this but I had left it in for a while.